libcontainer/init_linux.go

```diff
@@ -619,55 +599,14 @@ func signalAllProcesses(m cgroups.Manager, s os.Signal) error {
 		return err
 	}
 	for _, pid := range pids {
-		p, err := os.FindProcess(pid)
-		if err != nil {
-			logrus.Warn(err)
-			continue
-		}
-		procs = append(procs, p)
-		if err := p.Signal(s); err != nil {
-			logrus.Warn(err)
+		err := unix.Kill(pid, s)
+		if err != nil && err != unix.ESRCH { //nolint:errorlint // unix errors are bare
+			logrus.Warnf("kill %d: %v", pid, err)
 		}
 	}
 	if err := m.Freeze(configs.Thawed); err != nil {
 		logrus.Warn(err)
 	}
 
-	subreaper, err := system.GetSubreaper()
-	if err != nil {
-		// The error here means that PR_GET_CHILD_SUBREAPER is not
-		// supported because this code might run on a kernel older
-		// than 3.4. We don't want to throw an error in that case,
-		// and we simplify things, considering there is no subreaper
-		// set.
-		subreaper = 0
-	}
-
-	for _, p := range procs {
-		if s != unix.SIGKILL {
-			if ok, err := isWaitable(p.Pid); err != nil {
-				if !errors.Is(err, unix.ECHILD) {
-					logrus.Warn("signalAllProcesses: ", p.Pid, err)
-				}
-				continue
-			} else if !ok {
-				// Not ready to report so don't wait
-				continue
-			}
-		}
-
-		// In case a subreaper has been setup, this code must not
-		// wait for the process. Otherwise, we cannot be sure the
-		// current process will be reaped by the subreaper, while
-		// the subreaper might be waiting for this process in order
-		// to retrieve its exit code.
-		if subreaper == 0 {
-			if _, err := p.Wait(); err != nil {
-				if !errors.Is(err, unix.ECHILD) {
-					logrus.Warn("wait: ", err)
-				}
-			}
-		}
-	}
```

Comment:
As discussed on Matrix, this change makes sense but we should let folks know about it. The only folks I know who use libcontainer directly to manage containers fully is @LittleLightLittleFire.

@LittleLightLittleFire just letting you know that with this change, you will need to make sure you have a thread doing zombie reaping in whatever management process you're using libcontainer from (if you don't have that already).

Reply:
Yep that's fine, I think anyone that is using libcontainer would already use their initproc to reap, because otherwise zombies start piling up.

Reply:
@LittleLightLittleFire I found out that libcontainer implements own child reaper in `container.Signal`, which is useless for the  majority of the use cases and should not really belong here, so we're removing it (see the commit titled `libct: signalAllProcesses: remove child reaping`).

So, those users who did not have a proper child reaper, and use shared pidns for containers, might now see zombies everywhere as a result of this change (which we'll try to communicate in the release notes).

Reply:
> but we should let folks know about it

@cyphar I have added the proposed changelog entries to this PR description -- feel free to comment on the text, or just change it directly.